src/org/traccar/protocol/OwnTracksProtocolDecoder.java

```diff
@@ -0,0 +1,132 @@
+
+import java.net.SocketAddress;
+import java.nio.charset.StandardCharsets;
+import java.util.Date;
+
+import java.io.StringReader;
+import javax.json.Json;
+import javax.json.JsonObject;
+
+// import org.traccar.helper.Log;
+
+public class OwnTracksProtocolDecoder extends BaseProtocolDecoder {
+
+    public OwnTracksProtocolDecoder(OwnTracksProtocol protocol) {
+        super(protocol);
+    }
+
+    private void sendResponse(Channel channel, HttpResponseStatus status) {
+        if (channel != null) {
+            HttpResponse response = new DefaultHttpResponse(HttpVersion.HTTP_1_1, status);
+            response.headers().add(HttpHeaders.Names.CONTENT_LENGTH, 0);
+            channel.write(response);
+        }
+    }
+
+    @Override
+    protected Object decode(
+            Channel channel, SocketAddress remoteAddress, Object msg) throws Exception {
+
+        HttpRequest request = (HttpRequest) msg;
+        JsonObject root = Json.createReader(
+                        new StringReader(request.getContent().toString(StandardCharsets.US_ASCII))).readObject();
+
+        // Log.debug("**********JPJPJP " + root.getString("_type"));
+        Position position = new Position();
+        position.setProtocol(getProtocolName());
+        position.setValid(true);
+
+        position.setLatitude(root.getJsonNumber("lat").doubleValue());
+        position.setLongitude(root.getJsonNumber("lon").doubleValue());
+        // position.setValid(true);   // OwnTracks publishes only valid payloads
+
+        try {
+                position.setSpeed(UnitsConverter.knotsFromCps(root.getInt("vel")));
+        } catch (Exception e) { }
+
+        try {
+                position.setAltitude(root.getInt("alt"));
+        } catch (Exception e) { }
+
+        try {
+                position.setCourse(root.getInt("cog"));
+        } catch (Exception e) { }
+
+        try {
+                position.setAccuracy(root.getInt("acc"));
+        } catch (Exception e) { }
+
+        /*
+         * TID will be deviceId for now
+        try {
+                position.set("tid", root.getString("tid"));
+        } catch (Exception e) { }
+        */
+
+        try {
+                position.set("t", root.getString("t"));
+        } catch (Exception e) { }
+
+        try {
+                position.set(Position.KEY_BATTERY, root.getInt("batt"));
+        } catch (Exception e) {
+        }
+
+        long timestamp = root.getJsonNumber("tst").longValue();
+        if (timestamp < Integer.MAX_VALUE) {
+            timestamp *= 1000;
+        }
+        position.setTime(new Date(timestamp));
+
+        DeviceSession deviceSession = getDeviceSession(channel, remoteAddress, root.getString("tid"));
+        if (deviceSession == null) {
+                sendResponse(channel, HttpResponseStatus.BAD_REQUEST);
+                return null;
+        }
+        position.setDeviceId(deviceSession.getDeviceId());
+
+
+
+        if (position.getFixTime() == null) {
```

Comment:
Is this really necessary? Are there cases when app sends data without timestamp? If yes, please provide unit tests for the case.